posts/inside-rust/2022-07-26-keyword-generics.md

```diff
@@ -0,0 +1,405 @@
+we want to briefly share why we've started this initiative, and share some
+insight on what we're about.
+
+[Oli]: https://github.com/oli-obk
+[Niko]: https://github.com/nikomatsakis
+[Yosh]: https://github.com/yoshuawuyts
+[kwi]: https://github.com/rust-lang/keyword-generics-initiative
+
+[^initiative]: Rust governance terminology can sometimes get confusing. An
+"initiative" in Rust parlance is different from a "working group" or "team".
+Initiatives are intentionally limited: they exist to explore, design, and
+implement specific pieces of work - and once that work comes to a close, the
+initiative will wind back down. This is different from, say, the lang team -
+which essentially carries a `'static` lifetime - and whose work (ominously) does
+not have a clearly defined end.
+
+## A missing kind of generic
+
+One of Rust's defining features is the ability to write functions which are
+_generic_ over their input types. That allows us to write a function once,
+leaving it up to the compiler to generate the right implementations for us.
+
+Rust allows you to be generic over types - it does not allow you to be generic over other things that are usually specified by keywords in Rust. For example, whether a function is async, whether a function can fail or not, whether a function is const or not, etc.
+
+The post ["What color is your function"][color] [^color] describes what happens
+when a language introduces async functions, but with no way to be generic over
+them:
+
+> I will take async-await over bare callbacks or futures any day of the week.
+> But we’re lying to ourselves if we think all of our troubles are gone. As soon
+> as you start trying to write higher-order functions, or reuse code, you’re
+> right back to realizing color is still there, bleeding all over your codebase.
+
+This isn't just limited to async though, it applies to all modifier keywords -
+including ones we may define in the future. So we're looking to fill that gap
+by exploring something we call "keyword generics" [^name]: the ability to be
+generic over keywords such as `const` and `async`.
+
+[color]: https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/
+[^color]: R. Nystrom, “What Color is Your Function?,” Feb. 01, 2015.
+https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/
+(accessed Apr. 06, 2022).
+
+[^name]: The longer, more specific name would be: "keyword modifier generics".
+We've tried calling it that, but it's a bit of a mouthful. So we're just
+sticking with "keyword generics" for now, even if the name for this feature may
+end up being called something more specific in the reference and documentation.
+
+To give you a quick taste of what we're working on, this is roughly how we
+imagine you may be able to write a function which is generic over "asyncness"
+in the future:
+
+_Please note that this syntax is entirely made up, just so we can use something
```

Comment:
Would a footnote or `>` block look better?

Reply:
Changed it to use `>`!